pkg/kubefed2/federate/util.go

```diff
@@ -0,0 +1,127 @@
+*/
+
+package federate
+
+import (
+	"fmt"
+	"strings"
+
+	apiextv1b1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+)
+
+func CrdForAPIResource(apiResource metav1.APIResource) *apiextv1b1.CustomResourceDefinition {
+	scope := apiextv1b1.ClusterScoped
+	if apiResource.Namespaced {
+		scope = apiextv1b1.NamespaceScoped
+	}
+	return &apiextv1b1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: groupQualifiedName(apiResource),
+		},
+		Spec: apiextv1b1.CustomResourceDefinitionSpec{
+			Group:   apiResource.Group,
+			Version: apiResource.Version,
+			Scope:   scope,
+			Names: apiextv1b1.CustomResourceDefinitionNames{
+				Plural:   apiResource.Name,
+				Singular: apiResource.SingularName,
+				Kind:     apiResource.Kind,
+			},
+		},
+	}
+}
+
+func LookupAPIResource(config *rest.Config, key string) (*metav1.APIResource, error) {
+	client, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating discovery client: %v", err)
+	}
+
+	// TODO(marun) Allow the targeting of a specific group
+	// TODO(marun) Allow the targeting of a specific version
+
+	resourceLists, err := client.ServerPreferredResources()
+	if err != nil {
+		return nil, fmt.Errorf("Error listing api resources: %v", err)
+	}
+
+	// TODO(marun) Consider using a caching scheme ala kubectl
+	var targetResource *metav1.APIResource
+	for _, resourceList := range resourceLists {
+		for _, resource := range resourceList.APIResources {
+			if key == resource.Name ||
+				key == resource.SingularName ||
+				key == resource.Kind ||
+				key == strings.ToLower(resource.Kind) {
+
+				targetResource = &resource
+				break
+			}
+			for _, shortName := range resource.ShortNames {
+				if key == shortName {
+					targetResource = &resource
+					break
+				}
+			}
+			if targetResource != nil {
+				break
+			}
+		}
+		if targetResource != nil {
+			// The list holds the GroupVersion for its list of APIResources
+			gv, err := schema.ParseGroupVersion(resourceList.GroupVersion)
+			if err != nil {
+				return nil, fmt.Errorf("Error parsing GroupVersion: %v", err)
+			}
+			targetResource.Group = gv.Group
+			targetResource.Version = gv.Version
+			break
+		}
+	}
+
+	if targetResource != nil {
+		return targetResource, nil
+	}
+	return nil, fmt.Errorf("Unable to find api resource named %q.", key)
+}
+
+func resourceKey(apiResource metav1.APIResource) string {
+	var group string
+	if apiResource.Group == "" {
+		group = "core"
+	} else {
+		group = apiResource.Group
+	}
+	var version string
+	if apiResource.Version == "" {
+		version = "v1"
+	} else {
+		version = apiResource.Version
+	}
+	return fmt.Sprintf("%s.%s/%s", apiResource.Name, group, version)
+}
+
+func groupQualifiedName(apiResource metav1.APIResource) string {
+	if len(apiResource.Group) == 0 {
```

Comment:
(no action required) Just noticed we are using `len` to determine empty string and in the function right above we use `""`.